File: app/src/main/java/com/marked/vifo/activities/LoginActivity.java
package com.marked.vifo.activities;

import android.app.ProgressDialog;
import android.content.Intent;
import android.content.IntentFilter;
import android.graphics.Color;
import android.os.Bundle;
import android.support.design.widget.CoordinatorLayout;
import android.support.design.widget.Snackbar;
import android.support.v4.content.ContextCompat;
import android.support.v4.content.LocalBroadcastManager;
import android.support.v7.app.AppCompatActivity;
import android.view.View;
import android.widget.Button;
import android.widget.ImageView;

import com.marked.vifo.R;
import com.marked.vifo.extras.IHTTPStatusCodes;
import com.marked.vifo.gcm.RegistrationBroadcastReceiver;
import com.marked.vifo.gcm.extras.IgcmConstants;
import com.marked.vifo.gcm.services.LogInRegistrationIntentService;
import com.marked.vifo.helper.Utils;
import com.rengwuxian.materialedittext.MaterialEditText;

public class LoginActivity extends AppCompatActivity implements View.OnClickListener {
    private Button mLogInButtonPixy;
    private MaterialEditText mEmail, mPassword;
    private RegistrationBroadcastReceiver mRegistrationBroadcastReceiver;
    private ProgressDialog mProgressDialog;
    private CoordinatorLayout mContainer;
    private ImageView mMoreImageView;
    private LocalBroadcastManager mBroadcastManagerastManager;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_log_in);
        mLogInButtonPixy = (Button) findViewById(R.id.logInButtonPixy);
        mEmail = (MaterialEditText) findViewById(R.id.emailEditText);
        mPassword = (MaterialEditText) findViewById(R.id.passwordEditText);
        mContainer = (CoordinatorLayout) findViewById(R.id.container_log_in_fragment);
        mMoreImageView = (ImageView) findViewById(R.id.moreImageView);

        mLogInButtonPixy.setOnClickListener(this);
        mMoreImageView.setColorFilter(ContextCompat.getColor(this, R.color.primary));

        mBroadcastManagerastManager = LocalBroadcastManager.getInstance(this);

        // GCM registration
        mRegistrationBroadcastReceiver = new RegistrationBroadcastReceiver();
        mRegistrationBroadcastReceiver.setOnRegistrationBroadcastReceiverListener(new RegistrationBroadcastReceiver.RegistrationBroadcastReceiverListener() {
            @Override
            public void onDismiss() {
                mProgressDialog.dismiss();
            }

            @Override
            public void onError(int errorStatusCode) {
                Snackbar snackbar = null;
                switch (errorStatusCode) {
                    case IHTTPStatusCodes.REQUEST_CONFLICT:
                        snackbar = Utils.createWhiteSnackBar(LoginActivity.this, mContainer, "You already have an account");
                        break;
                    case IHTTPStatusCodes.REQUEST_TIMEOUT:
                        snackbar = Utils.createWhiteSnackBar(LoginActivity.this, mContainer, "Timeout error");
                        break;
                    case IHTTPStatusCodes.REQUEST_INCORRECT_PASSWORD:
                        snackbar = Utils.createWhiteSnackBar(LoginActivity.this, mContainer, "Incorrect password");
                        break;
                    case IHTTPStatusCodes.NOT_FOUND:
                        snackbar = Utils.createWhiteSnackBar(LoginActivity.this, mContainer, "We are sorry, but we did not found you");
                        break;
                }
                if (snackbar != null) {
                    snackbar.getView().setBackgroundColor(Color.WHITE);
                    snackbar.show();
                }
            }
        });

    }



    @Override
    public void onResume() {
        super.onResume();
        mBroadcastManagerastManager.registerReceiver(mRegistrationBroadcastReceiver, new IntentFilter(IgcmConstants.ACTION_LOGIN));
        mBroadcastManagerastManager.registerReceiver(mRegistrationBroadcastReceiver, new IntentFilter(IgcmConstants.ACTION_ERROR));
    }

    @Override
    protected void onPause() {
        LocalBroadcastManager.getInstance(this).unregisterReceiver(mRegistrationBroadcastReceiver);
        super.onPause();
    }

    @Override
    public void onClick(View v) {
        Intent intent = null;
        if (Utils.isOnline(this)) // if we got internet, process the click. else tell them to activate it
            switch (v.getId()) {
                case R.id.logInButtonPixy:
                    if (!Utils.checkEnteredData(this, mEmail.getText().toString(), mPassword.getText().toString(), mContainer))
                        break;
                    LogInRegistrationIntentService.startActionLogin(this, mEmail.getText().toString(), mPassword.getText().toString());
                    mProgressDialog = ProgressDialog.show(this, "Login", "Please wait ...", true);
                    break;
            }
        else
            Utils.showNoConnectionDialog(this);

    }
}

File: app/src/main/java/com/marked/vifo/ui/activity/LogInActivity.kt
package com.marked.vifo.ui.activity;

import android.app.ProgressDialog
import android.content.Intent
import android.content.IntentFilter
import android.os.Bundle
import android.support.v4.content.ContextCompat
import android.support.v4.content.LocalBroadcastManager
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.PopupMenu
import android.view.MenuItem
import android.view.View
import com.marked.vifo.R
import com.marked.vifo.delegate.DialogRegistration
import com.marked.vifo.extra.GCMConstants
import com.marked.vifo.gcm.RegistrationBroadcastReceiver
import com.marked.vifo.gcm.service.LogInRegistrationIntentService
import com.marked.vifo.helper.DataValidation
import com.marked.vifo.helper.Utils
import kotlinx.android.synthetic.main.activity_log_in.*

class LogInActivity : AppCompatActivity(), View.OnClickListener, PopupMenu.OnMenuItemClickListener {
	private val mProgressDialog by lazy { ProgressDialog(this) };
	private val mRegistrationBroadcastReceiver by lazy { RegistrationBroadcastReceiver(DialogRegistration(this, mProgressDialog)) }
	private val mBroadcastManagerastManager by lazy { LocalBroadcastManager.getInstance(this) };

	override
	protected fun onCreate(savedInstanceState: Bundle?) {
		super.onCreate(savedInstanceState);
		setContentView(R.layout.activity_log_in);
		moreImageView.setColorFilter(ContextCompat.getColor(this, R.color.primary));
	}

	override
	fun onResume() {
		super.onResume();
		mBroadcastManagerastManager.registerReceiver(mRegistrationBroadcastReceiver, IntentFilter(GCMConstants.ACTION_LOGIN));
		mBroadcastManagerastManager.registerReceiver(mRegistrationBroadcastReceiver, IntentFilter(GCMConstants.ACTION_ERROR));
	}

	override
	protected fun onPause() {
		mBroadcastManagerastManager.unregisterReceiver(mRegistrationBroadcastReceiver);
		super.onPause();
	}

	override
	fun onClick(v: View) {
		if (Utils.isOnline(this)) // if we got internet, process the click. else tell them to activate it
			when (v.id) {
				R.id.logInButtonPixy -> {
					if (!DataValidation(this, emailEditText.text.toString(), passwordEditText.text.toString()).validate())
						return;
					LogInRegistrationIntentService.startActionLogin(this, emailEditText.text.toString(), passwordEditText.text.toString());
					mProgressDialog.setTitle("Login")
					mProgressDialog.setMessage("Please wait...")
					mProgressDialog.isIndeterminate = true
					mProgressDialog.show()
				}
				R.id.signUpButton -> {
					startActivity(Intent(this, SignUpActivity::class.java))
				}
			}
		else
			Utils.showNoConnectionDialog(this);
	}

	override
	fun onMenuItemClick(item: MenuItem): Boolean {
		when (item.itemId) {
			R.id.aboutMenuItem -> {
				// TODO: 30-Nov-15 implement this
			}
		}
		return false;
	}

	/*
   * Triggered when the 'more' icon is clicked
   * The icon is in fragment_log_in.xml but handeled here because it won't get triggered in LogInFragment.java
   * */
	fun showPopup(view: View) {
		val popup = PopupMenu(this, view);
		val inflater = popup.menuInflater;
		inflater.inflate(R.menu.menu_popup_login, popup.menu);
		popup.setOnMenuItemClickListener(this);
		popup.show();
	}
}